metricbeat/module/system/diskio/diskstat_windows_helper.go

```diff
@@ -0,0 +1,169 @@
+
+	"github.com/shirou/gopsutil/disk"
+)
+
+const (
+	fileDeviceDisk               = 0x00000007
+	fileAnyAccess                = 0x0000
+	errorSuccess   syscall.Errno = 0
+)
+
+var (
+	// ioctlDiskPerformance :control code - https://docs.microsoft.com/en-us/windows/desktop/api/winioctl/ni-winioctl-ioctl_disk_performance
+	ioctlDiskPerformance        = getCtlCode(fileDeviceDisk, 0x0008, 0, fileAnyAccess)
+	modkernel32                 = syscall.NewLazyDLL("kernel32.dll")
+	procGetLogicalDriveStringsW = modkernel32.NewProc("GetLogicalDriveStringsW")
+)
+
+type logicalDrive struct {
+	Name    string
+	UNCPath string
+}
+
+type diskPerformance struct {
+	BytesRead           int64
+	BytesWritten        int64
+	ReadTime            int64
+	WriteTime           int64
+	IdleTime            int64
+	ReadCount           uint32
+	WriteCount          uint32
+	QueueDepth          uint32
+	SplitCount          uint32
+	QueryTime           int64
+	StorageDeviceNumber uint32
+	StorageManagerName  [8]uint16
+}
+
+// used to get the specific IOCTL_DISK_PERFORMANCE control code
+func getCtlCode(deviceType uint32, function uint32, method uint32, access uint32) uint32 {
+	return (deviceType << 16) | (access << 14) | (function << 2) | method
+}
+
+// IOCounters gets the diskio counters and maps them to the list of counterstat objects
+func iOCounters(names ...string) (map[string]disk.IOCountersStat, error) {
+	logicalDisks, err := getLogicalDriveStrings()
+	if err != nil || logicalDisks == nil {
+		return nil, err
+	}
+	ret := make(map[string]disk.IOCountersStat, 0)
+	for _, drive := range logicalDisks {
+		// not get _Total or Harddrive
+		if len(drive.Name) > 3 {
+			continue
+		}
+		//filter by included devices
+		if len(names) > 0 && !containsDrive(names, drive.Name) {
+			continue
+		}
+
+		var counter, err = iOCounter(drive.UNCPath)
+		if err != nil {
+			return nil, err
+		}
+		ret[drive.Name] = disk.IOCountersStat{
+			Name:       drive.Name,
+			ReadCount:  uint64(counter.ReadCount),
+			WriteCount: uint64(counter.WriteCount),
+			ReadBytes:  uint64(counter.BytesRead),
+			WriteBytes: uint64(counter.BytesWritten),
+			ReadTime:   uint64(counter.ReadTime),
+			WriteTime:  uint64(counter.WriteTime),
+		}
+	}
+	return ret, nil
+}
+
+// iOCounter calls syscal func CreateFile to generate a handler then executes the DeviceIoControl func in order to retrieve the metrics
+func iOCounter(path string) (diskPerformance, error) {
+	var diskPerformance diskPerformance
+	var diskPerformanceSize uint32
+	utfPath, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return diskPerformance, err
+	}
+	hFile, err := syscall.CreateFile(utfPath,
+		syscall.GENERIC_READ|syscall.GENERIC_WRITE,
+		syscall.FILE_SHARE_READ|syscall.FILE_SHARE_WRITE,
+		nil,
+		syscall.OPEN_EXISTING,
+		syscall.FILE_FLAG_BACKUP_SEMANTICS,
+		0)
+
+	if err != nil {
+		return diskPerformance, err
+	}
+	defer syscall.CloseHandle(hFile)
+
+	err = syscall.DeviceIoControl(hFile,
+		ioctlDiskPerformance,
```

Comment:
Should we return each of the devices back to their original state when the metricset stops? Like send `IOCTL_DISK_PERFORMANCE_OFF` if it was previously off.

Reply:
As previously discussed, for now it will be enabled as there could be other processes running on the machine that need this performance counter enabled.
As a somewhat workaround we are letting the users know in the logs that we are enabling these performance counters by adding the EnableCounterForIoctl  key in the windows registry.
The function `disablePerformanceCounters ` which will disable there perf counters has been added in the code for testing purposes only.